pkg/kubelet/kuberuntime/kuberuntime_container.go

```diff
@@ -19,12 +19,175 @@ package kuberuntime
 import (
 	"fmt"
 	"io"
+	"os"
+	"path"
 
+	"github.com/golang/glog"
 	"k8s.io/kubernetes/pkg/api"
+	runtimeApi "k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/runtime"
 	kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
+	"k8s.io/kubernetes/pkg/types"
 	"k8s.io/kubernetes/pkg/util/term"
 )
 
+// getContainerLogsPath gets log path for container.
+func getContainerLogsPath(containerName string, podUID types.UID) string {
+	return path.Join(podLogsRootDirectory, string(podUID), fmt.Sprintf("%s.log", containerName))
+}
+
+// generateContainerConfig generates container config for kubelet runtime api.
+func (m *kubeGenericRuntimeManager) generateContainerConfig(container *api.Container, pod *api.Pod, restartCount int, podIP string) (*runtimeApi.ContainerConfig, error) {
+	opts, err := m.runtimeHelper.GenerateRunContainerOptions(pod, container, podIP)
+	if err != nil {
+		return nil, err
+	}
+
+	_, containerName, cid := buildContainerName(pod, container)
```

Comment:
My point is such logic can live in the runtime shim code, e.g. the docker shim can construct the name with such way so nothing changed as today. and for rkt, it construct the annotation to include hash, namespace, etc.

When listing containers, the runtime should be able to return only containers that are managed by kubelet (e.g. docker list with the name filter, rkt list with the annotation filter), on top of that, runtimes can still apply whatever filters that kubelet passes to them.

I am generally against putting things into names, especially if kubelet forces all runtimes to do so. And as a result, we will have to maintain such naming scheme in the future to keep backwards compatibility, that's why I am saying this is an implicit contract.

Although I agree it could be somehow helpful for users to know what containers are managed by kubelet easier.